Source/SA/DPD/DPD.cs

```diff
@@ -0,0 +1,207 @@
+            specAnSignal.Dpd.Configuration.ConfigureLookupTableThreshold(selectorString, lutConfig.ThresholdEnabled, lutConfig.ThresholdLevel,
+                lutConfig.ThresholdType);
+            specAnSignal.Dpd.Configuration.ConfigureLookupTableStepSize(selectorString, lutConfig.LookupTableStepSize);
+        }
+
+        public static void ConfigureRFmxMP(RFmxSpecAnMX specAnSignal, MemoryPolynomialConfiguration mpConfig, string selectorString = "")
+        {
+            specAnSignal.Dpd.Configuration.ConfigureDpdModel(selectorString, RFmxSpecAnMXDpdModel.GeneralizedMemoryPolynomial);
+            specAnSignal.Dpd.Configuration.ConfigureMemoryPolynomial(selectorString, mpConfig.MemoryPolynomialOrder, mpConfig.MemoryPolynomialDepth);
+            specAnSignal.Dpd.Configuration.ConfigureGeneralizedMemoryPolynomialCrossTerms(selectorString, mpConfig.MemoryPolynomialLeadOrder,
+                mpConfig.MemoryPolynomialLagOrder, mpConfig.MemoryPolynomialLeadMemoryDepth, mpConfig.MemoryPolynomialLagMemoryDepth,
+                mpConfig.MemoryPolynomialMaxLead, mpConfig.MemoryPolynomialMaxLag);
+            specAnSignal.Dpd.Configuration.ConfigureIterativeDpdEnabled(selectorString, mpConfig.IterativeDpdEnabled);
+            specAnSignal.Dpd.ApplyDpd.ConfigureMemoryModelCorrectionType(selectorString, mpConfig.MemoryModelCorrectionType);
+        }
+        #endregion
+
+        #region ApplyDPD
+        public static LookupTableResults ApplyLookupTableDPD(RFmxSpecAnMX specAnSignal, NIRfsg rfsgSession, CommonConfiguration commonConfig, LookupTableConfiguration lutConfig, Waveform waveform, string selectorString = "")
+        {
+            LookupTableResults lutResults = new LookupTableResults();
+            IntPtr rfsgPtr;
+            RFmxSpecAnMXDpdApplyDpdIdleDurationPresent idlePresent;
+            idlePresent = waveform.IdleDurationPresent ? RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.True : RFmxSpecAnMXDpdApplyDpdIdleDurationPresent.False;
+            rfsgPtr = rfsgSession.GetInstrumentHandle().DangerousGetHandle();
+            specAnSignal.Initiate("", selectorString);
+            specAnSignal.Dpd.ApplyDpd.ConfigureLookupTableCorrectionType(selectorString, lutConfig.LookupTableCorrectionType);
+            specAnSignal.Dpd.ApplyDpd.ApplyDigitalPredistortion(selectorString, waveform.WaveformData, idlePresent, commonConfig.Timeout, ref lutResults.WaveformWithDpdComplexSingle,
+                out lutResults.PostDPDPAPR, out lutResults.PowerOffset);
+
+            waveform.WaveformData = lutResults.WaveformWithDpdComplexSingle;
+            waveform.SampleRate = 1 / lutResults.WaveformWithDpdComplexSingle.PrecisionTiming.SampleInterval.TotalSeconds;
+            waveform.PAPR_dB = lutResults.PostDPDPAPR;
+            waveform.SignalBandwidth_Hz = 0.8 * waveform.SampleRate;
+
+            SG.DownloadWaveform(rfsgSession, waveform);
+            rfsgSession.RF.PowerLevel = commonConfig.DutAverageInputPower - lutResults.PowerOffset;
```

Comment:
I think it would be better here to simply read back the existing power level rather than using the CommonConfig value for two reasons:
- If the user has modified the power level since the ConfigureCommon function they will end up with the wrong power level
- Combined with a change on line 153 we can avoid passing the CommonConfig struct to this function at all
```suggestion
            rfsgSession.RF.PowerLevel = rfsgSession.RF.PowerLevel - lutResults.PowerOffset;
```